### PR TITLE
Separate Gargoyle functions into new package

### PR DESCRIPTION
**Pull request by 3noch**
This is based on #16.

This allows us to avoid `gargoyle` but still use `rhyolite-backend-db`.

Motivations:
  1. Gargoyle is an excellent feature, but a feature we may not want in some apps. For example, docker-based deployments would rarely, if ever, want a built-in database inside the app container.
  1. Currently, `gargoyle-nix` has a very large nix closure. This could probably be improved, but for apps that don't want the feature anyway, we can easily reduce closure size by not having it.
  1. Gargoyle is not supported on macOS currently, so usage on mac is not useful and the closure size unnecessary.

I also moved `Backend.TH` to `Backend.DB.TH` ande re-exported it in `Backend.TH`. This is because the module is actually more generally useful and can be higher up in the dependency graph.

### CODE REVIEW

**Comment by 3noch** [comment]
cc @luigy

File: README.md [hunk 1]
```diff
@@ -41,63 +36,17 @@ obelisk.project ./. ({ pkgs, ... }:
...
-    rhyolite-backend = self.callCabal2nix "rhyolite-backend" (rhyolite-src + /backend) { websockets = self.websockets-obsidian; };
-
-    github = self.callCabal2nix "github" (pkgs.fetchFromGitHub {
-       owner="phadej";
-       repo="github";
-       rev="d1599c6eea1163eb97113e59f72081b88002b534";
-       sha256="0a8ndisqmfsf7gaw320ls0ffvk926q8pmcfihd31v25k9qm2i4bi";
-    }) {};
-  };
+    overrides = pkgs.lib.composeExtensions ((import rhyolite-src).lib args).haskellOverrides (self: super: {
```
Comment (luigy) [inline README.md]: `args` does not include `pkgs` which is a required attribute of taken by `lib`

File: datastructures/src/Rhyolite/SemiMap.hs [hunk 2]
```diff
@@ -75,8 +75,8 @@ instance (Ord k) => Monoid (SemiMap k v) where
 instance (Ord k) => Semigroup (SemiMap k v) where
   (<>) = mappend
 
-instance (ToJSON k, ToJSON v) => ToJSON (SemiMap k v)
-instance (Ord k, FromJSON k, FromJSON v) => FromJSON (SemiMap k v)
+instance (ToJSON k, ToJSON v, ToJSONKey k) => ToJSON (SemiMap k v)
```
Comment (luigy) [inline datastructures/src/Rhyolite/SemiMap.hs]:
```
common/src/Common/App.hs:313:10: error:
    • Could not deduce (ToJSONKey (Signed (AuthToken Identity)))
        arising from a use of ‘aeson-1.3.1.1:Data.Aeson.Types.ToJSON.$dmtoJSON’
      from the context: ToJSON a
        bound by the instance declaration
        at common/src/Common/App.hs:313:10-39
    • In the expression:
        aeson-1.3.1.1:Data.Aeson.Types.ToJSON.$dmtoJSON @(AppView a)
      In an equation for ‘toJSON’:
          toJSON
            = aeson-1.3.1.1:Data.Aeson.Types.ToJSON.$dmtoJSON @(AppView a)
      In the instance declaration for ‘ToJSON (AppView a)’
    |
313 | instance ToJSON a => ToJSON (AppView a)
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
common/src/Common/App.hs:314:10: error:
    • Could not deduce (FromJSONKey (Signed (AuthToken Identity)))
        arising from a use of ‘aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON’
      from the context: FromJSON a
        bound by the instance declaration
        at common/src/Common/App.hs:314:10-43
    • In the expression:
        aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON @(AppView a)
      In an equation for ‘parseJSON’:
          parseJSON
            = aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON @(AppView a)
      In the instance declaration for ‘FromJSON (AppView a)’
    |
314 | instance FromJSON a => FromJSON (AppView a)
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

File: datastructures/src/Rhyolite/SemiMap.hs [hunk 3]
```diff
@@ -75,8 +75,8 @@ instance (Ord k) => Monoid (SemiMap k v) where
...
+instance (Ord k, FromJSON k, FromJSON v, FromJSONKey k) => FromJSON (SemiMap k v)
```
Comment (luigy) [inline datastructures/src/Rhyolite/SemiMap.hs]:
```
common/src/Common/App.hs:314:10: error:
    • Could not deduce (FromJSONKey (Signed (AuthToken Identity)))
        arising from a use of ‘aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON’
      from the context: FromJSON a
        bound by the instance declaration
        at common/src/Common/App.hs:314:10-43
    • In the expression:
        aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON @(AppView a)
      In an equation for ‘parseJSON’:
          parseJSON
            = aeson-1.3.1.1:Data.Aeson.Types.FromJSON.$dmparseJSON @(AppView a)
      In the instance declaration for ‘FromJSON (AppView a)’
    |
314 | instance FromJSON a => FromJSON (AppView a)
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by luigy** [comment]
err I meant to post the comments in #16

File: datastructures/src/Rhyolite/SemiMap.hs (hunk 2 above)
Comment (luigy) [reply datastructures/src/Rhyolite/SemiMap.hs]: covered on https://github.com/obsidiansystems/rhyolite/commit/b905fa34f5cdafe5f5f5eaf7e20bfff5be664297

File: datastructures/src/Rhyolite/SemiMap.hs (hunk 3 above)
Comment (luigy) [reply datastructures/src/Rhyolite/SemiMap.hs]: covered on https://github.com/obsidiansystems/rhyolite/commit/b905fa34f5cdafe5f5f5eaf7e20bfff5be664297

**Comment by ali-abrar** [comment]
We should be using gargoyle wherever we can. Can we reduce the closure size if that's a concern?

**Comment by 3noch** [comment]
@ali-abrar We probably could, but even if we could I still think having gargoyle as a dependency of all `backend` is too presumptive. It only uses one function from gargoyle...

**Comment by 3noch** [comment]
Sorry misspoke, `backend-db`.

**Comment by ali-abrar** [comment]
Thanks, merged!

**Pull request merged**